kernel/src/process_checking.rs

```diff
@@ -0,0 +1,357 @@
+//! Traits, types, and sample implementations of application credentials
+//! checkers, used to decide whether an application can be loaded. See
+//| the [AppID TRD](../../doc/reference/trd-appid.md).
+
+use crate::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
+use crate::hil::digest::{ClientData, ClientHash, ClientVerify};
+use crate::hil::digest::{DigestDataVerify, Sha256};
+use crate::process::{Process, ShortID, State};
+use crate::utilities::cells::OptionalCell;
+use crate::utilities::cells::TakeCell;
+use crate::utilities::leasable_buffer::{LeasableBuffer, LeasableMutableBuffer};
+use crate::ErrorCode;
+use tock_tbf::types::TbfFooterV2Credentials;
+use tock_tbf::types::TbfFooterV2CredentialsType;
+
+/// What a AppCredentialsChecker decided a particular applications credential
+/// indicates about the runnability of an application binary.
+#[derive(Debug)]
+pub enum CheckResult {
+    /// Accept the credential and run the binary.
+    Accept,
+    /// Go to the next credential or in the case of the last one fall
+    /// back to the default policy.
+    Pass,
+    /// Reject the credential and do not run the binary.
+    Reject,
+}
+
+/// Receives callbacks on whether a credential was accepted or not.
+pub trait Client<'a> {
+    fn check_done(
+        &self,
+        result: Result<CheckResult, ErrorCode>,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    );
+}
+
+/// Implements a Credentials Checking Policy.
+pub trait AppCredentialsChecker<'a> {
+    fn set_client(&self, client: &'a dyn Client<'a>);
+    fn require_credentials(&self) -> bool;
+
+    fn check_credentials(
+        &self,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    ) -> Result<(), (ErrorCode, TbfFooterV2Credentials, &'a [u8])>;
+}
+
+/// Whether two processes have the same Application Identifier; two
+/// processes with the same Application Identifier cannot run concurrently.
+pub trait AppIdentification {
```

Comment:
```suggestion
pub trait AppUniqueness {
```
?

Reply:
Makes sense -- when I came up with the trait name  I thought `different_identifier` would be the major method, but in practice it's the implementation of `has_unique_identifier`.